text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,284 @@
+dbg!(foo); // foo is still 1, but the copy of `foo` in `closure` is 2
+```
+
+Note that `foo` is _copied_ during move in this example
+as `i32` implements `Copy`.
+
+If a closure captures multiple bindings,
+the `move` keyword makes them all captured by moving.
+To move only specific bindings,
+list them in parentheses after `move`:
+
+```rs
+let foo = 1;
+let mut bar = 2;
+let mut closure = move(mut foo) || {
+    foo += 10;
+    bar += 10;
+};
+closure();
+dbg!(foo, bar); // foo = 1, bar = 12
+```
+
+Note that the outer `foo` no longer requires `mut`;
+it is relocated to the closure since it defines a new binding.
+
+Moved bindings may also be renamed:
+
+```rs
+let mut foo = 1;
+let mut closure = move(mut bar = foo) || {
+    foo = 2;
+    bar = 3;
+};
+closure();
+dbg!(foo); // the outer `foo` is 2 as it was captured by reference
+```
+
+Bindings may be transformed when moved:
+
+```rs
+let foo = vec![1];
+let mut closure = move(mut foo = foo.clone()) || {
+    foo.push(2);
+};
+closure();
+dbg!(foo); // the outer `foo` is still [1] because only the cloned copy was mutated
+```
+
+The above may be simplified to `move(mut foo.clone())` as well.
+This simplification is only allowed
+when the transformation expression is a method call on the captured binding.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A closure expression has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ClosureExpression_ :\
+> &nbsp;&nbsp; ( `move` _MoveBindings_<sup>?</sup> )<sup>?</sup>\
+> &nbsp;&nbsp; ( `||` | `|` _ClosureParameters_<sup>?</sup> `|`  )\
+> &nbsp;&nbsp; (_Expression_ | `->` _TypeNoBounds_&nbsp;_BlockExpression_)>\
+> _MoveBindings_ :\
+> &nbsp;&nbsp; `(` ( _MoveBinding_ (`,` _MoveBinding_)<sup>\*</sup> `,`<sup>?</sup> )<sup>?</sup> `)`\
+> _MoveBinding_ :\
+> &nbsp;&nbsp; _NamedMoveBinding_ | _UnnamedMoveBinding_\
+> _NamedMoveBinding_ :\
+> &nbsp;&nbsp; _PatternNoTopAlt_ `=` _Expression_\
+> _UnnamedMoveBinding_ :\
+> &nbsp;&nbsp; `mut`<sup>?</sup> ( _IdentifierExpression_ | _MethodCallExpression_ )\
+> _ClosureParameters_ :\
+> &nbsp;&nbsp; _ClosureParam_ (`,` _ClosureParam_)<sup>\*</sup> `,`<sup>?</sup>\
+> _ClosureParam_ :\
+> &nbsp;&nbsp; _OuterAttribute_<sup>\*</sup> _PatternNoTopAlt_&nbsp;( `:` _Type_  )<sup>?</sup>
+
+Closure expressions are classified into two main types,
+namely _ImplicitReference_ and _ImplicitMove_.
+A closure expression is _ImplicitMove_ IF AND ONLY IF
+it starts with a `move` token immediately followed by a `|` token,
+without any parentheses in between.
+
+## _ImplicitReference_ closures
+
+When the parentheses for _MoveBindings_ is present, or when the `move` keyword is absent,
+the closure expression is of the _ImplicitReference_ type, where
+all local variables in the closure construction scope not shadowed by any _MoveBinding_
+are implicitly captured into the closure by shared or mutable reference on demand,
+preferring shared reference if possible.
```

Comment:
I'd like to ask for a special sigil for this "and other implicit captures" behavior:

```rs
move(x, y, ..) || { … }
```

Indeed, by requiring a trailing `..` in order to allow implicit captures:
  - such implicit capture is now clearer: the `x, y` enumeration is now visibly non-exhaustive;
  - it gives the option not to provide `, ..`, thereby requiring exhaustive enumeration of captures: _this gives the power to the person defining the closure to **restrict** the kind of captures a closure may use_.
      - Given that the motivation for this syntax is finer-grained control, it would be a pity not to be featuring such a capability, and have to wait for a later edition-requiring amendment to get into this behavior.
      - reasons for restricting external captures can be:
          - soundness of more advanced designs (_e.g._, macros and advanced lifetime shenanigans);
          - especially in FFI or other type-erased scenarios where certain lifetimes may have been lost;
          - (or when worried about things like `{A,}Rc` cycles in case of implicit by-move captures)
          - better control of a closure size;
          - _helps teachability of how closures work_: `move(x) |…| { … }` would always be an `x`-capturing closure, no matter the body of `...`, thereby properly decoupling implementation details from API surface (of the closure type).
  - it mirror struct pattern destructuring, wherein a `Foo { x, .. }` pattern is not expected to be exhaustive over the fields it has, whereas `Foo { x }` is.

Reply:
> Indeed, by requiring a trailing `..` in order to allow implicit captures:

I assume you mean implicit captures _by-value_? or keeping the current default capture behavior (without `move`)? `Foo { x, .. }` means all fields except `x` are ignored so it would make sense if all fields other than `x` were ignored in `move(x, ..)` when doing move captures

Reply:
+1 for this

If `..` also allowed to specify a default capture mode it would be a killer feature for UI development and other similar cases where a lot of variables are captured by cloning.

> `Foo { x, .. }` means all fields except x are ignored

IMO `..` just means "the rest" when used as a pattern, as it always matches all the elements that weren't already in some other way. It just so happens that with structs/tuples it doesn't allow to create bindings for the matches elements, though with slices this is possible (e.g. `[head, tail @ ..]`).